src/services/encryptionService.js

```diff
@@ -1,38 +1,170 @@
+import localforage from "localforage";
+import navigateTo from "@utils/navigation/navigate_to";
+import showCustomAlert from "@utils/show_alert";
+import store from "@store/store";
+import { insertChat } from "@store/values/Conversations";
+import { setSelectedConversation } from "@store/values/SelectedConversation";
+import { addUsers, upsertUser } from "@store/values/Participants";
 
 class EncryptionService {
+  #encryptionSession = null;
+  #account = null;
+  #chatIdAfterRegister = null;
+  #vodozemacInitialized = false;
+
   constructor() {
-    initVodozemac().then(() => {
-      console.log("[encryption] Vodozemac init Ok");
-    });
+    this.#initializeVodozemac();
+  }
+
+  async #initializeVodozemac() {
+    try {
+      await initVodozemac();
+      this.#vodozemacInitialized = true;
+      console.log("[encryption] Vodozemac initialized successfully");
+    } catch (error) {
+      console.error("[encryption] Failed to initialize Vodozemac", error);
+      showCustomAlert(
+        "Failed to initialize encryption. Please reload the app."
+      );
+    }
+  }
 
-    this.account = null;
+  setChatIdAfterRegister(id) {
+    this.#chatIdAfterRegister = id;
   }
 
-  #getAccount() {
-    if (this.account) {
-      return this.account;
+  validateIsUserAuth() {
+    return !!this.#account;
+  }
+
+  async #getAccount(password) {
+    if (this.#account) {
+      return this.#account;
     }
 
-    this.account = new Account();
+    const key = (await EncodeText(password)).slice(0, 32);
+    const encAuthKey = await localforage.getItem("account");
 
-    return this.account;
+    if (encAuthKey) {
+      try {
+        this.#account = Account.from_pickle(encAuthKey, key);
+      } catch (error) {
+        await localforage.removeItem("account");
+        console.error("[encryption] Account session expired", error);
+        showCustomAlert("Account session has expired. Please log in again.");
+        return null;
+      }
+    } else {
+      this.#account = new Account();
+      const encryptedKey = this.#account.pickle(key);
+      localforage.setItem("account", encryptedKey);
+    }
+    return this.#account;
   }
 
-  async registerDevice() {
-    const user = this.#getAccount();
-    user.generate_one_time_keys(50);
+  async logout() {
+    this.#account = null;
+  }
+
+  async registerDevice(password) {
+    const user = await this.#getAccount(password);
+    if (!user) return;
 
-    setEncryptedUser(user);
+    user.generate_one_time_keys(50);
+    console.log("Encryption: User account generated", user);
 
     const data = {
       identity_key: user.curve25519_key,
       signed_key: user.ed25519_key,
       one_time_pre_keys: Object.fromEntries(user.one_time_keys.entries()),
     };
-    await api.encryptedDeviceCreate(data);
+
+    try {
+      await api.encryptedDeviceCreate(data);
+      if (this.#chatIdAfterRegister) {
+        navigateTo(`/#${this.#chatIdAfterRegister}`);
```

Comment:
I would remove the navigation code from this service.
Let's have encrypted service responsible for encryption only.

Reply:
I suggest to remove all thes code from here:

```
     navigateTo(`/#${this.#chatIdAfterRegister}`);
        store.dispatch(
          setSelectedConversation({ id: this.#chatIdAfterRegister })
        );
```

and do this navigation from component or so